warehouse/transformer/safe.go

```diff
@@ -0,0 +1,273 @@
+
+	"github.com/iancoleman/strcase"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/warehouse/transformer/internal/reservedkeywords"
+	"github.com/rudderlabs/rudder-server/warehouse/transformer/internal/response"
+	"github.com/rudderlabs/rudder-server/warehouse/transformer/internal/snakecase"
+	"github.com/rudderlabs/rudder-server/warehouse/transformer/internal/utils"
+	whutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+var (
+	reLeadingUnderscores           = regexp.MustCompile(`^_*`)
+	reNonAlphanumericOrDollar      = regexp.MustCompile(`[^a-zA-Z0-9\\$]`)
+	reStartsWithLetterOrUnderscore = regexp.MustCompile(`^[a-zA-Z_].*`)
+)
+
+// safeNamespace returns a safe namespace for the given destination type and input namespace.
+// The namespace is transformed by removing special characters, converting to snake case,
+// and ensuring its safe (not starting with a digit, not empty, and not a reserved keyword).
+func safeNamespace(conf *config.Config, destType, input string) string {
+	namespace := strings.Join(extractAlphanumericValues(input), "_")
+
+	if !shouldSkipSnakeCasing(conf, destType) {
+		namespace = strcase.ToSnake(namespace)
+	}
+	if startsWithDigit(namespace) {
+		namespace = "_" + namespace
+	}
+	if namespace == "" {
+		namespace = "stringempty"
+	}
+	if reservedkeywords.IsNamespace(destType, namespace) {
+		namespace = "_" + namespace
+	}
+	return misc.TruncateStr(namespace, 127)
+}
+
+func extractAlphanumericValues(input string) []string {
+	var (
+		extractedValues []string
+		currentValue    strings.Builder
+	)
+
+	for _, c := range input {
+		if isAlphaAlphanumeric(c) {
+			currentValue.WriteRune(c)
+		} else if currentValue.Len() > 0 {
+			extractedValues = append(extractedValues, currentValue.String())
+			currentValue.Reset()
+		}
+	}
+	if currentValue.Len() > 0 {
+		extractedValues = append(extractedValues, currentValue.String())
+	}
+	return extractedValues
+}
+
+func isAlphaAlphanumeric(c int32) bool {
+	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
+}
+
+func shouldSkipSnakeCasing(conf *config.Config, destType string) bool {
+	configKey := fmt.Sprintf("Warehouse.%s.skipNamespaceSnakeCasing", whutils.WHDestNameMap[destType])
+	return conf.GetBool(configKey, false)
+}
+
+func safeTableNameCached(tec *transformEventContext, key string) (string, error) {
+	var tableName string
+	var err error
+
+	cacheKey := safeTableNameCacheKey{
+		destType:                     tec.event.Metadata.DestinationType,
+		tableName:                    key,
+		skipReservedKeywordsEscaping: tec.intrOpts.skipReservedKeywordsEscaping,
+	}
+	columnNameCached, ok := tec.cache.safeTableNameCache.Load(cacheKey)
+	if !ok {
+		tableName, err = safeTableName(tec.event.Metadata.DestinationType, tec.intrOpts, key)
+		if err != nil {
+			return "", err
+		}
+		tec.cache.safeTableNameCache.Store(cacheKey, tableName)
```

Comment:
What is the benefit of creating a cache ? Does the "transform" function involve heavy computation?

Reply:
Yes, it's not very heavy computation, but it still involves a couple of string manipulations, which might be crucial if the string is large. For a particular key since these don't changes, it should be fine to have a cache here.